Add support for next_query_flagIN

ls-michielrensen opened the pull request:
##### SUMMARY
This PR adds support for `next_query_flagIN` in the proxysql_query_rules plugin

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
`proxysql_query_rules`

##### ADDITIONAL INFORMATION
`next_query_flagIN` allows to chain query rules which is useful for when example you want to allow a SELECT after an INSERT/UPDATE/DELETE to run on the same hostgroup (avoiding replication lag).

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
- name: Add insert query rule
   proxysql_query_rules:
      match_digest: "^INSERT"
      destination_hostgroup: 1,
      next_query_flagIN: 1

- name: Add update query rule
   proxysql_query_rules:
      match_digest: "^UPDATE"
      destination_hostgroup: 1,
      next_query_flagIN: 1

- name: Add delete query rules
   proxysql_query_rules:
      match_digest: "^DELETE"
      destination_hostgroup: 1,
      next_query_flagIN: 1

- name: Add insert query rules
   proxysql_query_rules:
      match_digest: ".*"
      destination_hostgroup: 1,
      next_query_flagIN: 1
      comment: Match every queries after an INSERT/UPDATE/DELETE query
```

markuman commented on plugins/modules/proxysql_query_rules.py:
```diff
@@ -75,6 +75,13 @@
+        With C(CASELESS) the match is case insensitive. With C(GLOBAL) the replace
+        is global (replaces all matches and not just the first).
+        For backward compatibility, only C(CASELESS) is the enabled by default.
+    version_added: "2.9"
```
version added would be `1.3.0`.  
can you also please add a changelog fragement @ls-michielrensen?

futhermore I'm afraid that this will break backwards compatibility with proxysql 1.x.x like #71 #73

ls-michielrensen replied:
Yeah it probably is not backwards compatible no, do we need to add a version constraint somewhere?

markuman replied:
basically you can adopt like this: https://github.com/ansible-collections/community.proxysql/pull/73/commits/ca47c3218adbefbc8fd150e8e93d50cf71e3915c  
This is imo sufficient. The integration test must be updated/prepared. I'm working on that already in #73

Andersson007 replied:
link to the changelog related doc for convenience https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment

Andersson007 reviewed (commented):
@ls-michielrensen thanks for the PR!

I'm abstain from evaluating this changes as I'm not a specialist.

@markuman knows better:)

I would also add:
1. Example to the EXAMPLES block
2. Integration tests
3. Changelog [fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) as @markuman spotted
4. Mention in the options descriptions that they are supported since ProxySQL version ..., otherwise they should be ignored or the module should fail. Not sure which approach is better

ls-michielrensen commented:
> @ls-michielrensen thanks for the PR!
> 
> I'm abstain from evaluating this changes as I'm not a specialist.
> 
> @markuman knows better:)
> 
> I would also add:
> 
> 1. Example to the EXAMPLES blocks
Added the example from this PR description to the block
> 2. Integration tests
Uh gotta figure that out (first-timer here...)
> 3. Changelog [fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) as @markuman spotted
Changelog fragment added
> 4. Mention in the options descriptions that they are supported since ProxySQL version ..., otherwise they should be ignored or the module should fail. Not sure which approach is better
Added version constraint as suggested by @markuman in https://github.com/ansible-collections/community.proxysql/commit/ca47c3218adbefbc8fd150e8e93d50cf71e3915c

markuman commented on plugins/modules/proxysql_query_rules.py:
```diff
@@ -355,23 +368,29 @@ def __init__(self, module):
                             "replace_pattern",
                             "destination_hostgroup",
                             "cache_ttl",
-                            "cache_empty_result",
```
`cache_empty_result` is not needed to handle in this PL...can be reverted.

markuman replied:
Ok, blame me.

`next_query_flagIN` and `next_query_flagIN` both exists also in 1.4.15.

ls-michielrensen replied:
Ah alright, I'll leave it then. But if next_query_flagIN already existed this version constraint is not required then I guess

ls-michielrensen commented:
@markuman Your suggestion to add version constraints seemed to have broken the Integration tests. I guess we'll have to figure that out some other way...

markuman commented:
> @markuman Your suggestion to add version constraints seemed to have broken the Integration tests. I guess we'll have to figure that out some other way...

No, there is something failing that is not affected by your PR 

> TASK [test_proxysql_replication_hostgroups : verify idempotent - change for proxysql >= 2] ***

However, as [mentioned above](https://github.com/ansible-collections/community.proxysql/pull/74#discussion_r713214415), you can revert it, because the table also exists in 1.4.15. It was my fault.

markuman commented:
I see two sanity erros locally

```
ERROR: plugins/modules/proxysql_query_rules.py:268:82: W291: trailing whitespace
ERROR: plugins/modules/proxysql_query_rules.py:273:24: invalid-examples: EXAMPLES is not valid YAML
```

ls-michielrensen commented:
@markuman thanks, fixed the sanity failures, silly mistake on my end. Also reverted the version constraint 👍🏻

markuman commented:
Integration test pass locally for `test_proxysql_query_rules`.  
Last missing peace is to append the existing integration test and test the new parameters (_add/modify in check_mode, not in check_mode and idempotent_).

ls-michielrensen commented:
@markuman thanks for double checking. Can you point me in the right direction as to how to run the integration tests locally?

Andersson007 commented:
@ls-michielrensen you could use this tutorial https://github.com/ansible/community-docs/blob/main/create_pr_quick_start_guide.rst#test-your-changes

ls-michielrensen commented:
@Andersson007 thanks for the headsup, worked flawlessly 👍🏻 
@markuman I amended the integration tests to include `next_query_flagIN`. The tests are passing for me locally 👍🏻

markuman reviewed (approved):
LGTM

Andersson007 reviewed (approved):
LGTM from general perspective:)

Andersson007 commented:
@ls-michielrensen thanks for the contribution! @markuman thanks for reviewing and merging!

ls-michielrensen commented:
@Andersson007 you're welcome, was a fun experience and quickly executed 👍🏻
